### PR TITLE
extract.default using broom

### PR DESCRIPTION
**Pull request by vincentarelbundock**
**_I promise I will let you breath after this one!_**

``broom`` is a package designed to extract information on coefficient estimates and gof measures from models. It is part of ``tidyverse``, so it is highly visible and well-resourced. https://github.com/tidyverse/broom

``broom`` duplicates much of the functionality in your 4881 lines-long ``extract.R`` file. It covers many of the same models, but also supports several classes that you do not.

This pull request creates an ``extract.default`` method. When ``texreg`` does not offer a specialized extract method for a given model class, ``extract.default`` tries to use ``broom`` to extract the relevant info and feed it to ``createTexreg``. If ``broom`` is not installed, it recommends that users install it.

This immediately gives you support for several new models (I have not counted, but I'm guessing maybe 10 of them), as well as all future updates to the ``broom`` package. If you can live with dependencies, this could eventually allow you to delegate some of the extraction to the ``tidyverse`` folks.

Here's an example with 3 new models and 1 old one (you'll note that GOFs can become a bit unwieldy, but that's better than no support at all).

```
library(pacman)
p_load(broom, robust, lfe, texreg)
url = 'https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/HistData/Guerry.csv'
dat = read.csv(url, stringsAsFactors=FALSE)
dat$binary = ifelse(dat$Literacy > 30, 1, 0)
mod = list()
mod[['glm']] = glm(binary  ~ Lottery, family=binomial(), data=dat)
mod[['glmRob']] = glmRob(binary ~ Lottery, family=binomial(), data=dat)
mod[['felm']] = felm(Literacy ~ Lottery | Region, data=dat)
mod[['nls']] = nls(Literacy ~ a + b * Lottery, data=dat, start = c('a'=0, 'b'=0))
screenreg(mod)

===============================================================================
                       glm            glmRob         felm       nls
-------------------------------------------------------------------------------
(Intercept)              1.99074 ***    1.99074 ***
                        (0.54966)      (0.54966)
Lottery                 -0.03195 **    -0.03195 **   -0.09451
                        (0.01027)      (0.01027)     (0.05378)
a                                                                  50.28892 ***
                                                                   (3.53836)
b                                                                  -0.25363 ***
                                                                   (0.07065)
-------------------------------------------------------------------------------
AIC                    106.47952                                  727.73208
BIC                    111.38821                                  735.09513
Log Likelihood         -51.23976
Deviance               102.47952
Num. obs.               86                           85
deviance                              102.47952                 22219.01573
null.deviance                         119.22132
df.residual                            84.00000                    84.00000
R^2 (full model)                                      0.60528
R^2 (proj model)                                      0.03762
Adj. R^2 (full model)                                 0.58030
Adj. R^2 (proj model)                                -0.02329
sigma                                                              16.26383
isConv                                                              1.00000
finTol                                                              0.00000
logLik                                                           -360.86604
===============================================================================
*** p < 0.001, ** p < 0.01, * p < 0.05
```

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -7,7 +7,7 @@ Maintainer: Philip Leifeld <philip.leifeld@glasgow.ac.uk>
 Description: Converts coefficients, standard errors, significance stars, and goodness-of-fit statistics of statistical models into LaTeX tables or HTML tables/MS Word documents or to nicely formatted screen output for the R console for easy model comparison. A list of several models can be combined in a single table. The output is highly customizable. New model types can be easily implemented.
 URL: http://github.com/leifeld/texreg/
 BugReports: https://github.com/leifeld/texreg/issues
-Suggests: nlme, survival, network, ergm, lme4 (>= 1.0), btergm, Zelig (>= 5.0-16)
+Suggests: nlme, survival, network, ergm, lme4 (>= 1.0), btergm, broom, Zelig (>= 5.0-16)
```
Comment (leifeld) [inline DESCRIPTION]: As of recently, CRAN encourages indicating the minimum version number that is supported in brackets. Perhaps just throw in the current `broom` version number?

File: R/extract.R [hunk 2]
```diff
@@ -7,6 +7,33 @@
 setGeneric("extract", function(model, ...) standardGeneric("extract"), 
     package = "texreg")
 
+# default extract method prompts users to install the broom package
+extract.default <- function(model, ...) {
+  if (!'broom' %in% row.names(installed.packages())) {
+    stop("texreg does not directly support models of class ",
+         class(model), 
+         ", but it can sometimes use the ``broom`` library to extract model information. Call texreg again after installing the broom library to see if this is possible.")
```
Comment (leifeld) [inline R/extract.R]: Can you replace `broom library` by `broom package` (twice in that row)?

File: R/extract.R [hunk 3]
```diff
@@ -7,6 +7,33 @@
...
+      !all(c('term', 'estimate', 'std.error', 'p.value') %in% colnames(estimates))) { # tidy sometimes produces "non-standard" results
+    stop('Neither texreg nor broom supports models of class ', class(model), '.')
+  }
+  tr <- createTexreg(coef.names = estimates$term, 
+                     coef = estimates$estimate, 
+                     se = estimates$std.error, 
+                     pvalues = estimates$p.value, 
+                     gof.names = colnames(gof), 
+                     gof = as.numeric(gof[1, ]), 
+                     gof.decimal = rep(TRUE, ncol(gof))
```
Comment (leifeld) [inline R/extract.R]: Is it possible to find out if the respective GOF statistic is an integer variable?

File: R/extract.R [hunk 4]
```diff
@@ -7,6 +7,33 @@
...
+      (class(gof) == 'try-error') ||
+      !all(c('term', 'estimate', 'std.error', 'p.value') %in% colnames(estimates))) { # tidy sometimes produces "non-standard" results
+    stop('Neither texreg nor broom supports models of class ', class(model), '.')
+  }
+  tr <- createTexreg(coef.names = estimates$term, 
+                     coef = estimates$estimate, 
+                     se = estimates$std.error, 
+                     pvalues = estimates$p.value, 
+                     gof.names = colnames(gof), 
+                     gof = as.numeric(gof[1, ]),
```
Comment (leifeld) [inline R/extract.R]: What happens if the GOF statistic is not numeric and cannot be cast to a numeric variable? I suppose there will be an error message? Can you catch the error and omit the statistic in that case?

File: R/extract.R [hunk 5]
```diff
@@ -7,6 +7,33 @@
 setGeneric("extract", function(model, ...) standardGeneric("extract"), 
     package = "texreg")
 
+# default extract method prompts users to install the broom package
+extract.default <- function(model, ...) {
```
Comment (leifeld) [inline R/extract.R]: I am not familiar with `default` class names. Is this what is executed if the class name is not recognized? Or does `broom` really call its objects `default`?

File: R/extract.R (hunk 3 above)
Comment (vincentarelbundock) [inline R/extract.R]: I can't think of a robust way to do this (there probably is). A statistic which *looks* like an integer in one model may have stuff after the dot in a different model. I don't think ``broom`` is super careful about classes in the ``glance`` function.

File: R/extract.R (hunk 4 above)
Comment (vincentarelbundock) [inline R/extract.R]: I outsourced the gof extraction to ``broom_gof`` in ``internals.R``. Now, I'm putting everything in a data.frame, coercing using ``as.numeric``, and applying ``na.omit`` to the data.frame. There may be a more elegant solution, but this works.

File: R/extract.R (hunk 2 above)
Comment (vincentarelbundock) [inline R/extract.R]: Done

File: R/extract.R (hunk 5 above)
Comment (vincentarelbundock) [inline R/extract.R]: It's a fall back. For example:

```
> k <- NULL
> extract(k)
Error in extract(k) : 
  Neither texreg nor broom supports models of class NULL.
```

File: DESCRIPTION (hunk 1 above)
Comment (vincentarelbundock) [inline DESCRIPTION]: Done

**Comment by vincentarelbundock** [comment]
I'm aware that the docs still fail to build. Will let you know when I know why.

**Comment by vincentarelbundock** [comment]
Note: ``extract.default`` turned out to be a bad programming habit I carried over from S3. I moved it to an ``extract.broom`` method with signature "ANY".

Question: Do you have any idea why I get this error with the docs?

```
* checking Rd cross-references ... OK
* checking for missing documentation entries ... WARNING
Undocumented S4 methods:
  generic 'extract' and siglist 'ANY'
All user-level objects in a package (including S4 classes and methods)
should have documentation entries.
See chapter ‘Writing R documentation files’ in the ‘Writing R
Extensions’ manual.
```

Comparing my ``extract.broom`` to ``extract.Arima``, for example, it seems like they both have entries in all the same docs. I can't figure this one out.

File: R/extract.R (hunk 3 above)
Comment (leifeld) [inline R/extract.R]: The `gof.decimal` slot does not have to be identical across models for the same model term. See the following example:
```
library("texreg")

# example from ?lm
ctl <- c(4.17,5.58,5.18,6.11,4.50,4.61,5.17,4.53,5.33,5.14)
trt <- c(4.81,4.17,4.41,3.59,5.87,3.83,6.03,4.89,4.32,4.69)
group <- gl(2, 10, 20, labels = c("Ctl","Trt"))
weight <- c(ctl, trt)
lm.D9 <- lm(weight ~ group)
lm.D90 <- lm(weight ~ group - 1) # omitting intercept

model1 <- extract(lm.D9)
model2 <- extract(lm.D90)
model2@gof.decimal[3] <- TRUE

screenreg(list(model1, model2))
```
which produces this table:
```
=================================
             Model 1    Model 2  
---------------------------------
(Intercept)   5.03 ***           
             (0.22)              
groupTrt     -0.37       4.66 ***
             (0.31)     (0.22)   
groupCtl                 5.03 ***
                        (0.22)   
---------------------------------
R^2           0.07       0.98    
Adj. R^2      0.02       0.98    
Num. obs.    20         20.00    
RMSE          0.70       0.70    
=================================
*** p < 0.001, ** p < 0.01, * p < 0.05
```
Note the different decimal settings for `Num. obs`.

So if you can check in each case if the entry is integer, that should account for exceptions in other models.

File: DESCRIPTION [hunk 6]
```diff
@@ -12,3 +12,4 @@ Depends: R (>= 2.15.0)
 Imports: methods, grDevices, graphics, stats
 Enhances: AER, betareg, brglm, censReg, dynlm, eha, erer, fGarch, gamlss, gee, geepack, gmm, h2o, latentnet, lmtest, lqmm, MASS, mfx, mgcv, mlogit, mnlogit, MuMIn, nnet, ordinal, plm, pscl, quantreg, relevent, rms, robustbase, RSiena, sampleSelection, simex, sna, spdep, survey, systemfit, tergm, VGAM, xergm
 License: GPL-2 | GPL-3
+RoxygenNote: 6.0.1
```
Comment (leifeld) [inline DESCRIPTION]: We are not using roxygen. Why is there a roxygen note?

File: NAMESPACE [hunk 7]
```diff
@@ -16,6 +16,7 @@ export("coeftostring")
 export("createTexreg")
 exportMethods("extract")
 S3method("print", "texregTable")
+export("extract.broom")
```
Comment (leifeld) [inline NAMESPACE]: Can you please ensure that the `extract` methods are listed in the correct alphabetical order?

File: R/internal.R [hunk 8]
```diff
@@ -1308,3 +1308,41 @@ customcolumnnames <- function(modelnames, custom.columns, custom.col.pos,
...
+                'sigma' = 'Sigma',
+                'statistic' = 'Statistic'
+                )
+  gof_dict <- gof_dict[names(gof_dict) %in% out$gof.names]
+  idx <- match(names(gof_dict), out$gof.names)
+  out$gof.names[idx] <- gof_dict
+  # output
+  out <- stats::na.omit(out)
+  return(out)
+}
```
Comment (leifeld) [inline R/internal.R]: What is the advantage of having this `broom`-specific code in `internal.R` rather than in the `extract` method?

**Comment by leifeld** [comment]
About the build error: It looks like you wrote documentation for `broom` objects. But they don't exist. You ought to write documentation for `ANY` objects. E.g., `\alias{extract, ANY-method}` instead of `\alias{extract, broom-method}` and so on).

File: DESCRIPTION (hunk 6 above)
Comment (vincentarelbundock) [inline DESCRIPTION]: probably added by a ``devtools`` function I ran.  now removed.

File: NAMESPACE (hunk 7 above)
Comment (vincentarelbundock) [inline NAMESPACE]: ok

File: R/internal.R (hunk 8 above)
Comment (vincentarelbundock) [inline R/internal.R]: (a) less clutter in ``texreg.R``; (b) ``broom_gof`` is already pretty long and could eventually be extended further (e.g., better ``gof_dict`` to cleanup more ``gof.names`` automatically, and smarter type inference); (c) easier error catching at the level of the ``extract.ANY`` method.

**Comment by vincentarelbundock** [comment]
I added some minimal type inference to set better digits option. ``integer`` and ``logical`` both get ``digits = FALSE`` (so logical gets coerced to "1" or "0"). 

All checks have passed, so it looks like the S4 doc fix for ANY-method worked.

Output looks like this on my computer:

```
> library(pacman)
> p_load(broom, robust, lfe, texreg)
> url = 'https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/HistData/Guerry.csv'
> dat = read.csv(url, stringsAsFactors=FALSE)
> dat$binary = ifelse(dat$Literacy > 30, 1, 0)
> mod = list()
> mod[['glm']] = glm(binary  ~ Lottery, family=binomial(), data=dat)
> mod[['glmRob']] = glmRob(binary ~ Lottery, family=binomial(), data=dat)
> mod[['felm']] = felm(Literacy ~ Lottery | Region, data=dat)
> mod[['nls']] = nls(Literacy ~ a + b * Lottery, data=dat, start = c('a'=0, 'b'=0))
> screenreg(mod)

===================================================================
                       glm         glmRob      felm    nls         
-------------------------------------------------------------------
(Intercept)              1.99 ***    1.99 ***                      
                        (0.55)      (0.55)                         
Lottery                 -0.03 **    -0.03 **   -0.09               
                        (0.01)      (0.01)     (0.05)              
a                                                         50.29 ***
                                                          (3.54)   
b                                                         -0.25 ***
                                                          (0.07)   
-------------------------------------------------------------------
AIC                    106.48                            727.73    
BIC                    111.39                            735.10    
Log Likelihood         -51.24                           -360.87    
Deviance               102.48      102.48              22219.02    
Num. obs.               86                     85                  
Deviance (Null)                    119.22                          
DF Resid.                           84                    84       
R^2 (full model)                                0.61               
R^2 (proj model)                                0.04               
Adj. R^2 (full model)                           0.58               
Adj. R^2 (proj model)                          -0.02               
Sigma                                                     16.26    
Convergence                                                1       
Tolerance                                                  0.00    
===================================================================
*** p < 0.001, ** p < 0.01, * p < 0.05

```

**Comment by vincentarelbundock** [comment]
Note that ``texreg`` type inference relies on the ``broom::glance`` providing statistics in the correct type. I didn't try to do any guessing myself.

**Comment by leifeld** [comment]
About your last comment: It's OK that there is no guessing. But can you perhaps add a sensible error message if the `glance` output returns something weird that cannot be cast into `numeric`?

**Comment by vincentarelbundock** [comment]
I don't know of an object that produces non-numeric gof, so I don't have an easy way to test this, but I wrote a warning which should be printed when ``is.na(as.numeric(gof))``. https://github.com/vincentarelbundock/texreg/blob/cb8299116987185a8b857f004cf061fb25c67be2/R/internal.R#L1347

**Comment by leifeld** [comment]
Excellent! Thanks again for your contribution!

**Pull request merged**